Upgrade to Mono 5.10.x

dellis1972 opened the pull request:
Fixed an issue with the Xml Serializer expecting mcs when
serializing data.

dellis1972 committed:
Upgrade to Mono 4.4.0
Fixed an issue with the Xml Serializer expecting mcs when
serializing data.
Added Mkbundled mcs for OSX. Just incase mono decides to
fallback on Temp assemblies during serialization.

For some reason the environment variables

MONO_XMLSERIALIZER_THS=no
MONO_REFLECTION_SERIALIZER=yes

do not seem to work.

dellis1972 commented:
@mrhelmut @cra0zy this is working fine for me on Mac and Linux now. :)

ThomasFOG commented:
Tested working. :+1: 

Do you know against which version of libc you built it?

dellis1972 commented:
@mrhelmut Hope this has the info you want.

```
/lib/i386-linux-gnu/libc.so.6
GNU C Library (Ubuntu GLIBC 2.23-0ubuntu3) stable release version 2.23, by Roland McGrath et al.
Copyright (C) 2016 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.
There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE.
Compiled by GNU CC version 5.3.1 20160413.
Available extensions:
        crypt add-on version 2.1 by Michael Glad and others
        GNU Libidn by Simon Josefsson
        Native POSIX Threads Library by Ulrich Drepper et al
        BIND-8.2.3-T5B
libc ABIs: UNIQUE IFUNC
For bug reporting instructions, please see:
<https://bugs.launchpad.net/ubuntu/+source/glibc/+bugs>.

```

dellis1972 commented:
I'm in the process of adding mkbundled versions of mcs.exe this is to get around Serialisation issues where mono falls back to using mcs to compile a reflection assembly. 

No idea why but the env vars below do not seem to stop it trying to use mcs.

MONO_XMLSERIALIZER_THS=no
MONO_REFLECTION_SERIALIZER=yes

dellis1972 commented:
Added mcs for Linux and Mac. Along with a wrapper script based on the 'Kick' one.

ThomasFOG commented:
Libc 2.23, this may be a bit high for Linux compatibility. Upstream is built with libc 2.15.
But I'm not a Linux expert and I don't know how much this is important.

dellis1972 commented:
so I'll need to get an older version of Linux to build this on then :(

On 7 July 2016 at 11:25, Thomas Altenburger notifications@github.com
wrote:

> Libc 2.23, this may be a bit high for Linux compatibility. Upstream is
> built with libc 2.15.
> But I'm not a Linux expert and I don't know how much this is important.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> https://github.com/MonoGame/MonoKickstart/pull/9#issuecomment-231041100,
> or mute the thread
> https://github.com/notifications/unsubscribe/AAxeeZfY6VhpKoNNUcRDRxmGYDRogyaUks5qTNQvgaJpZM4JDVkM
> .

dellis1972 commented:
ubuntu 14.04 has libc v2.19 .. is that going to be old enough? //cc @mrhelmut @cra0zy

ThomasFOG commented:
I don't know Linux enough to be sure.
Since 4.4 is now out, maybe @flibitijibibo & @urkle would have interest in updating upstream? (with your reflection fix)

flibitijibibo commented on kick.c:
```diff
@@ -120,6 +120,7 @@ int main (int argc, char* argv[])
 	}
 	free(exe);
 
+	newargs [k++] = "-O=-aot";
```
Shouldn't this crash...? The newargs needs one more char*, right?

dellis1972 replied:
it does.. but it doesn't crash, I'll be fixing that later.

flibitijibibo commented:
I'll probably update to 4.4 after the current FNA wave is done, keeping it on 2.15. Once 12.04 LTS is EOL'd next year I'll be bumping up all my requirements to 2.19.

For upstream, we probably won't roll in things like the AOT and MCS inclusions, and I'm not exactly sure what the new envs are for... do they mean anything for non-AOT?

dellis1972 commented:
@flibitijibibo the "mcs" stuff and -O=-aot fixes a few issues you will get in 4.4

1) One linux mono runtime WILL CRASH if you don't have -O=-aot . This is a known issue on linux and I spoke to the xamarin runtime team about it. This is the work around for now until they fix it in a later version. 

2) The mcs / env vars fix issues with the XmlSerializer.  Its to do with the way the mono Serializer sometimes falls back to compiling a "temp" assembly. Now the envvar's are supposed to fix it. But as a backup having mcs present makes sure everything works as expected.

@mrhelmut I'll download ubuntu 12.04 and update this PR with the appropriate binaries.

dellis1972 commented:
Mono 4.4 crash on Linux is
"Error: No managed allocator, but we need one for AOT.
Are you using non-standard GC options?"

which is what the `-O=-aot` fixes
reference : https://bugzilla.xamarin.com/show_bug.cgi?id=42058

flibitijibibo commented:
Oh dear, and for a moment I thought that Xamarin had started testing their Mono builds again. What I get for finding the one stable release of Mono since 2.10.9!

I'll have to look at the setenvs, Apotheon is one game that like to spit out errors about XML serialization but for whatever reason it continues to work... if you want to goof around with existing code, Apotheon and Cryptark are two games that emit related errors (but continue to work somehow).

dellis1972 committed:
Linux binaries built against Ubuntu 12.04

dellis1972 committed:
Updated with the latest linux binaries.

dellis1972 committed:
Fixed a possible memory issue with the newargs length.

dellis1972 committed:
Updated Linux mcs binaries to be linked against libc 2.15

harry-cpp commented:
Just gonna add that mkbundle path is unusable, since they broke it....

dellis1972 commented:
@cra0zy @mrhelmut I updated all the kick.bin.\* mcs.bin.\* files to be linked against libc 2.15. So we should be ok now.

dellis1972 committed:
Fixed execute permissions on mcs script

dellis1972 committed:
Fixed a pathing issue in the mcs script

dellis1972 committed:
Fix an issue running mcs when running under Steam

Steam sets the LD_PRELOAD environment variable to
force the loading of its overlay early on in the
load process.

The problem is they attempt to load both 32 and 64
bit overlays. Probably because they don't know if
the game is 32 or 64 bit. This results in `ld`
issuing errors (which it ignores) on the stderr.

When `mcs` is called by mono to compile a temp
assembly during serialization it checks the
stderr for issues. Even though the `ld` errors
are not related to `mcs` it still fails. The
result is that the serialization system thinks
the build failed.

This fix we clear out the LD_PRELOAD once we
start the app. We know they will have been loaded
already and this prevents `mcs` trying to load them
too.

dellis1972 committed:
Removed 'mcs' script in favour of using the binaries directly.
The 'Kick' script is now reponsible for setting up the correct
symlink on startup.

ThomasFOG commented:
I've been testing this on MacOS, and so far so good. It brings fixes to System.Net that are very welcome.

Mind the encoding of Kick & mcs, they are DOS encoded instead of using the Unix encoding, and this may fail on some CLI (or is it my git config messing around?).

dellis1972 commented:
@mrhelmut having some problems with mcs on certain linux machines. (Mint 17.03 for example). 
The mcs script will be removed in favour of symlinking directly to the correct mcs.bin.xxxx file.

dellis1972 committed:
Added config files for mono 4.5 to resolve issue with System.Diagnostics
Updated kick app to set the MONO_CFG_DIR evn variable so that we
do NOT pick up system mono config ever. If we do we end up crashing.
Updted mcs linux binaries to be compeletly self contained due to
native crashes on certain linux distro's

flibitijibibo commented on precompiled/Kick:
```diff
@@ -25,11 +25,14 @@ if [ "$UNAME" == "Darwin" ]; then
 	./kick.bin.osx $@
 else
 	if [ "$ARCH" == "x86_64" ]; then
+		ln -sf mcs.bin.x86_64 mcs
```
While this works on Steam, this could be a problem in the long term. In particular, GOG/MojoSetup builds will NOT be a fan of this - if a game is installed in a location not normally accessible by the user (i.e. non-root) then the ln call will fail. If the game really depends on `mcs`, it's probably safer to make 32/64-bit builds separate, with mcs being the native mkbundle binary renamed, skipping the need for symlinks. In AppAdmin this would simply be 2 depots limited to Linux32 and Linux64 (with actual game content in the existing Linux multiarch depot), containing the Kick binaries exclusive to those architectures, and that same layout can be applied to MojoSetup packages without runtime symlinking.

It sounds like a PITA, but so is this whole XML incompatibility... MONOOOOOOOOO :shakefist:

dellis1972 commented:
@mrhelmut do you know what the min OSX version you build the libSDL2 against?

ThomasFOG commented:
I didn't built the dylib myself, it is coming from @flibitijibibo 's latest FNA redist (http://fna-xna.github.io/download/). Snow Leopard 10.6.8?

(the only dll I built is the 64bit Windows, the other are FNA's)

flibitijibibo commented:
Yeah, currently it targets Snow Leopard.

dellis1972 committed:
Fixed the mono/config file to work on mac and linux

dellis1972 committed:
Upgrade OSX to mono 4.8.1
Did some work to allow Mono to pick up dll's from a MonoBundle directory on Mac
Did some work to allow Mono to get its config files from a directory in Resources

dellis1972 committed:
Update mscorlib to mono 4.8.1

dellis1972 committed:
Updated README

dellis1972 committed:
Updated Linux binaries to mono 4.8

dellis1972 commented:
Updated to mono 4.8 which is in RC

ThomasFOG commented:
I'll give this a spin in the coming days.

dellis1972 committed:
Updated Assemblies to mono 4.8

dellis1972 committed:
Got rid of the need to have symlinks for mac packaging

opcon commented:
I was running into SSL verification issues on Linux while using this branch to deploy my game, turns out Mono is in the process of switching to a new SSL provider (see http://tirania.org/blog/archive/2016/Sep-30.html, http://www.mono-project.com/docs/about-mono/releases/4.8.0/#tls-12-support) which is present in `kick.bin.osx` but absent from the `kick.bin.x86*` executables.

A workaround for now is to set the environment variable `MONO_TLS_PROVIDER` to `legacy` or `old` as shown [here](https://github.com/mono/mono/blob/0007e99a87229eab592c179430c4fe3d885e1363/mcs/class/Mono.Security/Mono.Security.Interface/MonoTlsProviderFactory.cs#L56) when running on linux, to force it to use the old provider.

Just thought I'd note this in case anyone else is looking for reasons why their https connections were failing.

(As a side note, does anyone have a good solution for using the system SSL certificate store when running through Kickstart? At the moment I have to disable certificate verification in my code, since Kickstart-ed executables can't find a certificate store, which is pretty dodgy.)

dellis1972 commented:
Hey Patrick

Thanks for the heads up. I'll take a look and see if I can figure out what
is going on with the new provider.

Regarding certs, I'll take a look, but it might be possible to put
certificates into the mono configuration directory somewhere.

On 23 February 2017 at 01:17, Patrick Yates <notifications@github.com>
wrote:

> I was running into SSL verification issues on Linux while using this
> branch to deploy my game, turns out Mono is in the process of switching to
> a new SSL provider (see http://tirania.org/blog/archive/2016/Sep-30.html,
> http://www.mono-project.com/docs/about-mono/releases/4.8.0/#tls-12-support)
> which is present in kick.bin.osx but absent from the kick.bin.x86*
> executables.
>
> A workaround for now is to set the environment variable MONO_TLS_PROVIDER
> to legacy or old as shown here
> <https://github.com/mono/mono/blob/0007e99a87229eab592c179430c4fe3d885e1363/mcs/class/Mono.Security/Mono.Security.Interface/MonoTlsProviderFactory.cs#L56>
> when running on linux, to force it to use the old provider.
>
> Just thought I'd note this in case anyone else is looking for reasons why
> their https connections were failing.
>
> (As a side note, does anyone have a good solution for using the system SSL
> certificate store when running through Kickstart? At the moment I have to
> disable certificate verification in my code, since Kickstart-ed executables
> can't find a certificate store, which is pretty dodgy.)
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/MonoGame/MonoKickstart/pull/9#issuecomment-281859187>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAxeecobbDRvqdO-QUgj24VJXCH63yiUks5rfN4jgaJpZM4JDVkM>
> .
>

dellis1972 committed:
updated osx bin

dellis1972 committed:
Updated to latest 4.8.0 Binaies

dellis1972 committed:
Updated script to build mac and linux

dellis1972 committed:
Added pkgconfig paths

dellis1972 committed:
fixed typo

dellis1972 committed:
Updated x86 linux bin

dellis1972 committed:
Fixed Mac Build Include path

dellis1972 committed:
Added FindMono.cmake

dellis1972 committed:
Changed the way we pick up the mono-2.pc

dellis1972 committed:
Updated .travis.yml

dellis1972 committed:
Opps missed the ..

dellis1972 committed:
Fixed up Include directories and removed 32bit support

dellis1972 committed:
Added export of PKG_CONFIG_PATH

dellis1972 committed:
Added some paths

dellis1972 committed:
Updated FindMono and CMakeLists.txt

dellis1972 committed:
added back universal builds for osx

dellis1972 committed:
mono 5.2 update for testing

dellis1972 committed:
Added script to upgrade mono binaries

dellis1972 committed:
Added .gitignore

dellis1972 committed:
Add 32Bit Ubuntu Support via docker

dellis1972 committed:
Updated to using Ubuntu12.04 via docker

dellis1972 committed:
'fix build order

dellis1972 committed:
Updated Mac Build

dellis1972 committed:
Update include path

dellis1972 committed:
Fix mono library name

dellis1972 committed:
Update binaries to mono 5.4 for a test

dellis1972 committed:
Switch to using mono-sgen

dellis1972 committed:
switch back to mono-2

dellis1972 committed:
Mono 5.2.0.224

dellis1972 committed:
Updated dll.config

dellis1972 committed:
Add Patch for the mcs build

dellis1972 committed:
Added building of mcs

dellis1972 committed:
Add Mac build for mcs

dellis1972 committed:
Fixed typo

dellis1972 committed:
Removed mac flags from linux build

dellis1972 committed:
Fixed build again

dellis1972 committed:
Execute buildmcs locally

dellis1972 committed:
Fixed mcs build again

dellis1972 committed:
Statically compile mcs

dellis1972 committed:
Updated binaries

dellis1972 committed:
Try to use xcode6.4

tomspilman commented:
@dellis1972 - What is the status of this PR?

dellis1972 commented:
@tomspilman I guess I need to update to mono 5.8 now its stable, which means I need to build all the binaries again.. sigh.
I think @mrhelmut was going to test some of this stuff out.

dellis1972 commented:
I'll have another go at getting travis-ci to upload the binaries to GitHub releases again

dellis1972 committed:
Try to only build one mac build

tomspilman commented:
@dellis1972 - Should i set this up on our TeamCity like our other dependencies?

dellis1972 committed:
Fixed names of linux bin files. Added Travis releases

dellis1972 committed:
Add Linux.zip to releases

dellis1972 commented:
@tomspilman it might be nice but we have some very specific requirements. The versions of Linux MUST be `Ubuntu-12.04`or a distro at that level. This is so the binaries will work on as many versions of linux as possible. If you look at the travis scripts you will see I am using `docker` to manage the building of linux.

dellis1972 committed:
fix .travis.yml

tomspilman commented:
> The versions of Linux MUST be `Ubuntu-12.04` or a distro at that level.

Well would this be the same distro we would want to use to build/test all MG stuff?  Because will have to work that out anyway.

dellis1972 committed:
fix .travis.yml

dellis1972 committed:
fix .travis.yml

dellis1972 committed:
fix .travis.yml

dellis1972 committed:
fix linux build scripts

dellis1972 committed:
Enabled releases on all branches

dellis1972 committed:
changes to linux scripts

dellis1972 committed:
Add some debug output

dellis1972 committed:
changes to .travis.yml

dellis1972 committed:
Fixed Mac Releases Hopefully

dellis1972 committed:
Fixed filenames for linux

dellis1972 committed:
Split Linux and Mac releases

dellis1972 committed:
Hopefully this will work

dellis1972 committed:
ok maybe this works

dellis1972 committed:
again

dellis1972 committed:
Added ARM Support

ThomasFOG commented on CMakeLists.txt:
```diff
@@ -27,7 +27,11 @@ elseif (CMAKE_SIZEOF_VOID_P MATCHES "8" AND NOT(FORCE32))
+      SET(CMAKE_EXECUTABLE_SUFFIX ".bin.arm")
+    else()
+      SET(CMAKE_EXECUTABLE_SUFFIX ".bin.x86")
+    endif()
```
@dellis1972 Did you have any chance to test it yet? :)

dellis1972 replied:
it worked for a simple console app on my Pi3. I've not tested a full MG app yet though.

dellis1972 committed:
Add precompiled to the zips

dellis1972 committed:
Updated Upgrade script

dellis1972 committed:
force and upgrade for travis builds

dellis1972 committed:
Fix upgrade for Linux

dellis1972 committed:
more fixes

dellis1972 committed:
Added System.IO.Compression

dellis1972 committed:
Test PR deployment

dellis1972 committed:
publish tags

dellis1972 committed:
Added code to get the zip paths

ThomasFOG commented:
Hey,

I was wondering if there is any way to make kickstart to use the ```x64``` & ```x86``` folders for dll discovery instead of the ```lib``` & ```lib64``` ones? That would match the recent dll loading mechanism updated for .Net Core in MonoGame.
Right now, we have to package games with both type of folders in the event of the use of third party native libraries (like Steamworks).

dellis1972 commented:
@mrhelmut all that will be controlled via the various dll.config files for those third party native libs (Steamworks). They need to be updated to look in the right folders.

dellis1972 committed:
Removed before_deploy

urkle commented:
@mrhelmut you should simply be able to update the RPATH that is set on the executable as that is what decides where to look for the packaged libs that the binary needs to startup.

dellis1972 committed:
Change file wildcard

dellis1972 committed:
Added some logging

dellis1972 committed:
Fixed logging

dellis1972 committed:
more changes

dellis1972 committed:
Added a change in path

dellis1972 committed:
Update mono

dellis1972 committed:
Include libMonoPosixHelper.*

dellis1972 committed:
Fixed mac build issue

ThomasFOG commented:
```Include libMonoPosixHelper```

That will definitely help a bunch of users who are struggling to figure out how the native dependencies work.

dellis1972 committed:
Add config for compression

dellis1972 committed:
corrected name of config file

dellis1972 committed:
Output mono version

dellis1972 committed:
copy over the linux mscorlib.dll

dellis1972 committed:
Try to specify the same mono version as linux docker uses

dellis1972 committed:
Try to specify the same mono version as linux docker uses

dellis1972 committed:
Fixed lookup for libMonoPosixHelper.so on Linux

dellis1972 committed:
Match the x86 and x64 folders we use in MonoGame

dellis1972 committed:
Update CMake to set BIN_LIBROOT and BIN_RPATH to the new x86 and x64 folders

dellis1972 commented:
@mrhelmut so I updated the lib and lib64 to be x86/x64 and updated the CMakeList.txt to reflect that change. So we can just have the one set of directories for native libraries.

dellis1972 commented:
The next problem. mono no longer supports ubuntu 12.04 so the latest mono we can get is `5.10.1.47` (I think). That results in a different mscorlib.dll from Mac which is now `5.12.0.*` This means separate packages for Linux and Mac, not sure how that will effect people using this. 

alternatively I can move this so that 14.04 is the min this build automatically for.

ThomasFOG commented:
Cool!
I believe that we can move forward. I imagine gamers on Linux to be on more recent versions, so jumping to a more recent LTS should be safe enough. 12.04 support has ended anyway.

ThomasFOG commented:
Hey @dellis1972 

I'm updating our kickstart to your latest release. Any chance to update the BIN_LIBROOT for OSX as well?
MonoGame looks for native libraries at the ```./``` folder, while kickstart defines it in the ```./osx``` folder.

dellis1972 commented:
Should be able to do that. looks like a one liner

On Tue, 4 Sep 2018 at 09:08, Thomas Altenburger <notifications@github.com>
wrote:

> Hey @dellis1972 <https://github.com/dellis1972>
>
> I'm updating our kickstart to your latest release. Any chance to update
> the BIN_LIBROOT for OSX as well?
> MonoGame looks for native libraries at the ./ folder, while kickstart
> defines it in the ./osx folder.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/MonoGame/MonoKickstart/pull/9#issuecomment-418279063>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAxeeWpT19-54PpfVVELwKho64BaXmG9ks5uXjTzgaJpZM4JDVkM>
> .
>

dellis1972 committed:
Rework OSX Bin Path and Arm Bin Path

dellis1972 committed:
Fixed typo

dellis1972 commented:
@mrhelmut the v0.0.11 tag should (in theory) do the trick. Let me know if it doesn't

dellis1972 commented:
@mrhelmut now that I think about it. It should always look in the same dir as the binary anyway, so perhaps that change was not needed...

ThomasFOG commented:
Thanks! This latest version is really convenient to use with MonoGame, it's as simple as a copy/paste.

You might want to update the ```Kick``` script as well for this line: ```export DYLD_LIBRARY_PATH=$DYLD_LIBRARY_PATH:./```

dellis1972 committed:
More fixup

ThomasFOG commented on precompiled/Kick:
```diff
@@ -12,7 +12,7 @@ ARCH=`uname -m`
 if [ "$UNAME" == "Darwin" ]; then
 	# ... Except on OSX.
-	export DYLD_LIBRARY_PATH=$DYLD_LIBRARY_PATH:./osx/
+	export DYLD_LIBRARY_PATH=$DYLD_LIBRARY_PATH:./:./osx/
```
@dellis1972 I believe it was the wrong insertion, it should be ```./``` not ```./:./osx/```. :)

ThomasFOG commented:
I've been running this build successfully for quite some time now.

I am running into an issue, though. I can't use ```WebClient``` for http or https requests. The runtime throws a ```NotSupportedException``` stating that ```The URI prefix is not recognized```.

I suspect that there is a problem with the mono config files.

There are two files with the same content:
- ```./monomachineconfig```
- ```./mono/4.5/machine.config```

Which one should be the correct one? From what I understand from those files, the http/https prefixes should be supported. Which makes me think that those files aren't loaded?

Any idea @dellis1972 @flibitijibibo ? (meh at how badly mono documents those files)

flibitijibibo commented:
`monomachineconfig` was my lazy way of getting a file in the top directory rather than have two folders for a lone file - if this string is in the executable, `monomachineconfig` is the correct one, otherwise it's `mono/4.5/machine.config`:

https://github.com/flibitijibibo/MonoKickstart/blob/master/kick.c#L147

Not sure about the error though. Usually I get hit by https needing to be validated with a special callback, so at the very least I know it _can_ work...

ThomasFOG commented:
Thanks for pointing that out, looks like this PR is expecting ```./mono/config``` (which would resolve to ```./mono/config/4.5/machine.config``` I guess?): https://github.com/MonoGame/MonoKickstart/blob/9d8ed4e129356065abae72f749cadb2801a75e71/kick.c#L145

I remember having WebClient working on your branch, so that's why I've been puzzled about this.

ThomasFOG commented:
Okay, it might be a MacOS thing, because the config file is correctly loaded from ```mono/4.5/machine.config``` on Linux (and WebClient is working there).

ThomasFOG commented:
Ahah... on MacOS, the ```./mono/config``` folder has to exist so that ```./mono/4.5/machine.config``` gets loaded.

Same goes for ```../Resources/mono/config```, if it exists, then ```../Resources/mono/4.5/machine.config``` will be loaded.

Also, side effect of loading the config through ```setenv("MONO_CFG_DIR", ...)``` is that mono will create temporary config folders (```mono/registry/LocalMachine```) in the ```mono``` folder, which may be blocked by MacOS or refused in the Mac App Store.

ThomasFOG commented:
For reference, I've setup a branch to update MonoKickstart for our games: https://github.com/mrhelmut/MonoKickstart

It basically is the same as flibitijibibo's fork with the MonoGame RPATH (and updates Mono to latest stable because 5.10 has a bug in the BCL that we needed to be fixed).

It lacks mac app store certification compliance and mcs support.
